### PR TITLE
Fnymmm/add token generator

### PR DESCRIPTION
**Pull request by fnymmm**
Added the ability to generate Tokens based on ClientId and SecretId.
The secret in plain text is being kept server side.

-Secret has to be generated with the "save secret" option enabled and be published.

### CODE REVIEW

File: src/Server/src/AuthTokenGenerator/Certification/CertificateDataProtector.cs [hunk 1]
```diff
@@ -0,0 +1,152 @@
+using System.Security.Cryptography;
+using System.Security.Cryptography.X509Certificates;
+using IdOps.Abstractions;
+using IdOps.Certifaction;
+using IdOps.Models;
+using Serilog;
+
+namespace IdOps.Certification
+{
+    public class CertificateDataProtector : IDataProtector
```
Comment (PascalSenn) [inline src/Server/src/AuthTokenGenerator/Certification/CertificateDataProtector.cs:10]: What do we need this for?

File: src/Server/src/AuthTokenGenerator/Certification/IDataProtector.cs [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+namespace IdOps.Certification
+{
+    public interface IDataProtector
+    {
+        string Name { get; }
+
+        void Setup(EncryptionKeySetting settings);
+        EncryptionKeySetting SetupNew();
+        byte[] UnProtect(byte[] data);
```
Comment (PascalSenn) [inline src/Server/src/AuthTokenGenerator/Certification/IDataProtector.cs:11]: Unprotect

File: src/Server/src/AuthTokenGenerator/DefaultShellService.cs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        {
+            _userSettingService = userSettingService;
+        }
+
+        public string GetDefault()
+        {
+            UserSettings settings = _userSettingService.GetAsync(default)
+                .GetAwaiter()
+                .GetResult();
+
```
Comment (PascalSenn) [inline src/Server/src/AuthTokenGenerator/DefaultShellService.cs:21]: Nope => make it async

File: src/Server/src/AuthTokenGenerator/Certification/SymmetricEncryption.cs [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+            return plainData;
+        }
+
+        private static SymmetricAlgorithm GetAlgorithm(string algorithm)
+        {
+            switch (algorithm.ToUpper())
+            {
+                case "AES256":
+                    return new AesManaged()
```
Comment (PascalSenn) [inline src/Server/src/AuthTokenGenerator/Certification/SymmetricEncryption.cs:68]: Isnt this deprecated?

File: src/Server/src/AuthTokenGenerator/DefaultShellService.cs [hunk 5]
```diff
@@ -0,0 +1,40 @@
+using System.Runtime.InteropServices;
+using IdOps.Abstractions;
+using IdOps.Models;
+
+namespace IdOps
+{
+    public class DefaultShellService : IDefaultShellService
```
Comment (PascalSenn) [inline src/Server/src/AuthTokenGenerator/DefaultShellService.cs:7]: What?

File: src/Server/src/AuthTokenGenerator/IdentityService.cs [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+        public async Task<RequestTokenResult> RequestTokenAsync(
+            TokenRequestData request,
+            CancellationToken cancellationToken)
+        {
+            DiscoveryDocumentResponse disco = await _httpClientWrapper.GetDiscoveryDocumentAsync(
+                request.Authority,
+                cancellationToken);
+            TokenResponse? response = null;
+
+            if (request.GrantType == "client_credentials")
```
Comment (PascalSenn) [inline src/Server/src/AuthTokenGenerator/IdentityService.cs:33]: Create methods for each GranType that you want to process here.

File: src/Server/src/GraphQL/Client/ClientMutations.cs [hunk 7]
```diff
@@ -27,18 +30,17 @@ public class ClientMutations
...
+
             Client client = await _clientService.UpdateClientAsync(input, cancellationToken);
 
             return new SaveClientPayload(client);
         }
 
         [AuthorizeClientAuthoring(AccessMode.Write, includeTenantAuth: false)]
-        public async Task<AddClientSecretPayload> AddClientSecretAsync(
-            AddClientSecretRequest input,
+        public async Task<AddClientSecretPayload> AddClientSecretAsync(AddClientSecretRequest input,
```
Comment (PascalSenn) [inline src/Server/src/GraphQL/Client/ClientMutations.cs:43]: Please do not change the formatting

**Review by PascalSenn — COMMENTED** [review]
I partially reviewed this.
But i dont understand a few things.

- What do we need asymetric encryption for?
- Why do we not just use the Dataprotector API if we really need it?
- Why do we store the tokens?
and a few more

File: src/Server/src/AuthTokenGenerator/IdentityService.cs [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+            {
+                response = await _httpClientWrapper.RequestClientCredentialsTokenAsync(
```
[the remark this answers is not in the document]
Comment (fnymmm) [reply src/Server/src/AuthTokenGenerator/IdentityService.cs:35]: I created this wrapper so i can mock it, httpClients use extention methods for these requests://

File: src/Server/src/GraphQL/Client/TokenRequestDataResultFactory.cs [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+        string secretEncrypted =
+            client.ClientSecrets.First(secret => secret.Id.Equals(input.SecretId)).EncryptedSecret;
+        var secretDecrypted =
+            await _encryptionService.DecryptAsync(secretEncrypted, cancellationToken);
+
+        var grantTypes = client.AllowedGrantTypes.First();
+
+        var scopeIds =
+            client.AllowedScopes.ToList().ConvertAll(clientScope => clientScope.Id );
```
Comment (PascalSenn) [inline src/Server/src/GraphQL/Client/TokenRequestDataResultFactory.cs:49]: client.AllowedScopes.Select(x => x.Id).ToList() ?

Comment (PascalSenn) [reply src/Server/src/GraphQL/Client/TokenRequestDataResultFactory.cs:49]: Maybe it doesnt even need to be a `ToList` when `scopeService` accepts a `IEnumerable<Guid>`

File: src/Server/src/AuthTokenGenerator/IdentityService.cs [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+            TokenRequestData request,
+            CancellationToken cancellationToken)
+        {
+            DiscoveryDocumentResponse disco = await _tokenClient.GetDiscoveryDocumentAsync(
+                request.Authority,
+                cancellationToken);
+
+            TokenResponse? response = request.GrantType switch
+            {
+                "client_credentials" => await RequestClientCredentialTokenAsync(request, disco),
```
Comment (PascalSenn) [inline src/Server/src/AuthTokenGenerator/IdentityService.cs:32]: What do we do with `authorization_code`? Do we already have a plan for this?

File: src/Server/src/GraphQL/Client/SaveClientPayload.cs [hunk 11]
```diff
@@ -1,3 +1,5 @@
+
```
Comment (PascalSenn) [inline src/Server/src/GraphQL/Client/SaveClientPayload.cs:1]: Empty lines

**Review by PascalSenn — CHANGES_REQUESTED** [review]
It looks much better now. I made a few comments.

I think we still carry too much from Boost into IdOps. For example, the TokenModel. Do we need all these properties? What does the UI need? If we pass it out through the GraphQL API, then the frontend should have a requirement for it. Otherwise, it's probably not necessary.

**Comment by fnymmm** [comment]
everything merged with other pr

**Pull request closed without merging**